File: examples/example-kubernetes-client-openapi-operator-python/src/main/python/micronaut/operator/ConfigMapReconcilerWithFilters.py
import logging

from micronaut.context.annotation import Requires
from micronaut.context.env import Environment
from micronaut.kubernetes.client.openapi.informer.handler import Informer
from micronaut.kubernetes.client.openapi.model import V1ConfigMap
from micronaut.kubernetes.client.openapi.operator import Operator, OperatorResourceLister
from micronaut.kubernetes.client.openapi.operator.controller.reconciler import Request, ResourceReconciler, Result

from micronaut.operator.OnAddFilter import OnAddFilter
from micronaut.operator.OnDeleteFilter import OnDeleteFilter
from micronaut.operator.OnUpdateFilter import OnUpdateFilter

LOG = logging.getLogger(__name__)


# Only one informer per resource type and namespace can be registered, so this reconciler is
# enabled instead of ConfigMapReconciler by setting the example.config-map-filters property
@Requires(env=Environment.KUBERNETES)
@Requires(property="example.config-map-filters", value="true")
# tag::operator[]
@Operator(informer=Informer(apiType=V1ConfigMap),
          onAddFilter=OnAddFilter,
          onUpdateFilter=OnUpdateFilter,
          onDeleteFilter=OnDeleteFilter)
class ConfigMapReconcilerWithFilters(ResourceReconciler[V1ConfigMap]):
    # end::operator[]

    def reconcile(self, request: Request, lister: OperatorResourceLister[V1ConfigMap]) -> Result:
        LOG.info("Reconciling annotated config map: %s", request)
        # .. reconcile
        return Result(False)
    # tag::operator[]
    ...
# end::operator[]

File: examples/example-kubernetes-client-openapi-operator-kotlin/src/main/kotlin/micronaut/operator/ConfigMapReconcilerWithFilters.kt
package micronaut.operator

import io.micronaut.context.annotation.Requires
import io.micronaut.context.env.Environment
import io.micronaut.kubernetes.client.openapi.informer.handler.Informer
import io.micronaut.kubernetes.client.openapi.model.V1ConfigMap
import io.micronaut.kubernetes.client.openapi.operator.Operator
import io.micronaut.kubernetes.client.openapi.operator.OperatorResourceLister
import io.micronaut.kubernetes.client.openapi.operator.controller.reconciler.Request
import io.micronaut.kubernetes.client.openapi.operator.controller.reconciler.ResourceReconciler
import io.micronaut.kubernetes.client.openapi.operator.controller.reconciler.Result
import org.slf4j.LoggerFactory

// Only one informer per resource type and namespace can be registered, so this reconciler is
// enabled instead of ConfigMapReconciler by setting the example.config-map-filters property
@Requires(env = [Environment.KUBERNETES])
@Requires(property = "example.config-map-filters", value = "true")
//tag::operator[]
@Operator(informer = Informer(apiType = V1ConfigMap::class),
        onAddFilter = OnAddFilter::class,
        onUpdateFilter = OnUpdateFilter::class,
        onDeleteFilter = OnDeleteFilter::class)
class ConfigMapReconcilerWithFilters : ResourceReconciler<V1ConfigMap> {
//end::operator[]

    companion object {
        private val LOG = LoggerFactory.getLogger(ConfigMapReconcilerWithFilters::class.java)
    }

    override fun reconcile(request: Request, lister: OperatorResourceLister<V1ConfigMap>): Result {
        LOG.info("Reconciling annotated config map: {}", request)
        // .. reconcile
        return Result(false)
    }
//tag::operator[]
}
//end::operator[]
